File: backend/app/services/budget_guard.py
"""Deterministic budget guardrails for generated plans and editable ledgers."""

from __future__ import annotations

import copy
import math
from typing import Any, Iterable

from ..models.schemas import BudgetLedgerItem, DayPlan, TripRequest


TRANSPORT_BUFFER_PER_PERSON = 50.0
HOTEL_BUFFER_PER_ROOM_NIGHT = 300.0
PENDING_BUFFER_PER_PERSON = {
    "attraction": 100.0,
    "meal": 80.0,
    "transport": TRANSPORT_BUFFER_PER_PERSON,
    "other": 100.0,
}
MEAL_FLOOR_PER_PERSON = {
    "breakfast": 15,
    "lunch": 30,
    "dinner": 40,
    "snack": 15,
}


def group_budget_limit(
    budget_amount: Any,
    budget_basis: str,
    traveler_count: int,
) -> float | None:
    """Normalize a user budget to the canonical group-total basis."""
    try:
        amount = float(budget_amount)
    except (TypeError, ValueError):
        return None
    if not math.isfinite(amount) or amount <= 0:
        return None
    travelers = max(1, int(traveler_count or 1))
    if budget_basis == "per_person":
        amount *= travelers
    return round(amount, 2)


def pending_budget_buffer(
    items: Iterable[BudgetLedgerItem],
    traveler_count: int,
) -> float:
    """Reserve money for unquoted rows without pretending it is a real price."""
    travelers = max(1, int(traveler_count or 1))
    total = 0.0
    for item in items:
        if item.deleted or item.amount is not None:
            continue
        if item.type == "hotel":
            total += (
                HOTEL_BUFFER_PER_ROOM_NIGHT
                * max(1, int(item.room_count or 1))
                * max(1, int(item.nights or 1))
            )
            continue
        total += PENDING_BUFFER_PER_PERSON.get(item.type, 100.0) * travelers
    return round(total, 2)


def calculate_budget_status(
    totals: dict[str, float],
    items: Iterable[BudgetLedgerItem],
    traveler_count: int,
    budget_amount: Any,
    budget_basis: str,
    *,
    adjustment_applied: bool = False,
    adjustment_note: str = "",
) -> dict[str, Any]:
    """Return auditable quoted and projected budget status values."""
    materialized = list(items)
    limit = group_budget_limit(budget_amount, budget_basis, traveler_count)
    quoted_total = round(float(totals.get("total", 0) or 0), 2)
    buffer_amount = pending_budget_buffer(materialized, traveler_count)
    projected_total = round(quoted_total + buffer_amount, 2)
    return {
        "budget_limit": limit,
        "quoted_total": quoted_total,
        "over_budget_amount": (
            round(max(0.0, quoted_total - limit), 2) if limit is not None else 0.0
        ),
        "pending_buffer": buffer_amount,
        "projected_total": projected_total,
        "projected_over_budget_amount": (
            round(max(0.0, projected_total - limit), 2)
            if limit is not None
            else 0.0
        ),
        "adjustment_applied": bool(adjustment_applied),
        "adjustment_note": str(adjustment_note or ""),
    }


def _plan_known_total(days: list[DayPlan], traveler_count: int, room_count: int) -> float:
    attractions = sum(
        max(0, attraction.ticket_price) * traveler_count
        for day in days
        for attraction in day.attractions
    )
    meals = sum(
        max(0, meal.estimated_cost) * traveler_count
        for day in days
        for meal in day.meals
    )
    hotels = sum(
        max(0, day.hotel.estimated_cost) * room_count
        for day in days
        if day.hotel and day.hotel.price_status != "unavailable"
    )
    return round(attractions + meals + hotels, 2)


def _plan_pending_buffer(days: list[DayPlan], traveler_count: int, room_count: int) -> float:
    transport_rows = sum(1 for day in days if day.transportation.strip())
    hotel_nights = sum(
        1
        for day in days
        if day.hotel and day.hotel.price_status == "unavailable"
    )
    return round(
        transport_rows * TRANSPORT_BUFFER_PER_PERSON * traveler_count
        + hotel_nights * HOTEL_BUFFER_PER_ROOM_NIGHT * room_count,
        2,
    )


def _adjustment_note(language: str, reduction: float, buffer_amount: float) -> str:
    code = (language or "zh").strip().lower().split("-")[0]
    if code == "en":
        return (
            f"Meal estimates were reduced by ¥{reduction:g} to respect the budget, "
            f"with ¥{buffer_amount:g} reserved for unquoted items."
        )
    if code == "ja":
        return (
            f"予算に合わせて食事の見積もりを ¥{reduction:g} 引き下げ、"
            f"未見積もり項目に ¥{buffer_amount:g} を確保しました。"
        )
    return (
        f"已根据总预算将餐饮预估下调 ¥{reduction:g}，"
        f"并为待报价项目预留 ¥{buffer_amount:g}。"
    )


def adjust_generated_days_to_budget(
    request: TripRequest,
    days: list[DayPlan],
) -> tuple[list[DayPlan], dict[str, Any]]:
    """Make one conservative adjustment pass over unverified meal estimates.

    Provider hotel prices and itinerary attractions are left untouched. If meal
    estimates cannot absorb the excess, the remaining overage is reported by
    the ledger status instead of fabricating cheaper prices.
    """
    adjusted = [copy.deepcopy(day) for day in days]
    travelers = max(1, request.traveler_count)
    rooms = max(1, request.room_count or (travelers + 1) // 2)
    limit = group_budget_limit(request.budget_amount, request.budget_basis, travelers)
    buffer_amount = _plan_pending_buffer(adjusted, travelers, rooms)
    known_total = _plan_known_total(adjusted, travelers, rooms)
    if limit is None or known_total + buffer_amount <= limit:
        return adjusted, {
            "budget_adjustment_applied": False,
            "budget_adjustment_note": "",
        }

    remaining_reduction = known_total + buffer_amount - limit
    original_total = known_total
    meals = [meal for day in adjusted for meal in day.meals]
    meals.sort(
        key=lambda meal: (
            meal.estimated_cost - MEAL_FLOOR_PER_PERSON.get(meal.type.lower(), 20),
            meal.estimated_cost,
        ),
        reverse=True,
    )
    for meal in meals:
        floor = MEAL_FLOOR_PER_PERSON.get(meal.type.lower(), 20)
        reducible = max(0, meal.estimated_cost - floor)
        if reducible <= 0 or remaining_reduction <= 0:
            continue
        reduction_per_person = min(
            reducible,
            int(math.ceil(remaining_reduction / travelers)),
        )
        meal.estimated_cost -= reduction_per_person
        remaining_reduction = max(
            0.0,
            remaining_reduction - reduction_per_person * travelers,
        )

    adjusted_total = _plan_known_total(adjusted, travelers, rooms)
    reduction = round(max(0.0, original_total - adjusted_total), 2)
    applied = reduction > 0
    return adjusted, {
        "budget_adjustment_applied": applied,
        "budget_adjustment_note": (
            _adjustment_note(request.language or "zh", reduction, buffer_amount)
            if applied
            else ""
        ),
    }

File: backend-ts/src/domain/budget-guard.ts
import type { BudgetLedgerItem, BudgetTotals } from "./budget-ledger.ts";
import type { DayPlan, TripPlanningRequest } from "./orchestrator.ts";

const TRANSPORT_BUFFER_PER_PERSON = 50;
const HOTEL_BUFFER_PER_ROOM_NIGHT = 300;
const PENDING_BUFFER_PER_PERSON = { attraction: 100, meal: 80, transport: 50, other: 100 } as const;
const MEAL_FLOOR_PER_PERSON: Record<string, number> = { breakfast: 15, lunch: 30, dinner: 40, snack: 15 };

function rounded(value: number): number {
  return Math.round(value * 100) / 100;
}

function positiveCount(value: unknown, fallback = 1): number {
  const numeric = Number(value);
  return Number.isFinite(numeric) ? Math.max(1, Math.trunc(numeric)) : fallback;
}

export function groupBudgetLimit(
  budgetAmount: unknown,
  budgetBasis: unknown,
  travelerCount: number,
): number | null {
  if (budgetAmount === null || budgetAmount === undefined || budgetAmount === "") return null;
  const amount = Number(budgetAmount);
  if (!Number.isFinite(amount) || amount <= 0) return null;
  return rounded(amount * (budgetBasis === "per_person" ? positiveCount(travelerCount) : 1));
}

export function pendingBudgetBuffer(items: BudgetLedgerItem[], travelerCount: number): number {
  const travelers = positiveCount(travelerCount);
  return rounded(items.reduce((total, item) => {
    if (item.deleted || item.amount !== null) return total;
    if (item.type === "hotel") {
      return total + HOTEL_BUFFER_PER_ROOM_NIGHT
        * positiveCount(item.room_count)
        * positiveCount(item.nights);
    }
    return total + (PENDING_BUFFER_PER_PERSON[item.type as keyof typeof PENDING_BUFFER_PER_PERSON] ?? 100)
      * travelers;
  }, 0));
}

export function calculateBudgetStatus(
  totals: Pick<BudgetTotals, "total">,
  items: BudgetLedgerItem[],
  travelerCount: number,
  budgetAmount: unknown,
  budgetBasis: unknown,
  adjustment = { applied: false, note: "" },
): Record<string, number | boolean | string | null> {
  const limit = groupBudgetLimit(budgetAmount, budgetBasis, travelerCount);
  const quotedTotal = rounded(Number(totals.total || 0));
  const pendingBuffer = pendingBudgetBuffer(items, travelerCount);
  const projectedTotal = rounded(quotedTotal + pendingBuffer);
  return {
    budget_limit: limit,
    quoted_total: quotedTotal,
    over_budget_amount: limit === null ? 0 : rounded(Math.max(0, quotedTotal - limit)),
    pending_buffer: pendingBuffer,
    projected_total: projectedTotal,
    projected_over_budget_amount: limit === null ? 0 : rounded(Math.max(0, projectedTotal - limit)),
    adjustment_applied: adjustment.applied,
    adjustment_note: adjustment.note,
  };
}

function knownTotal(days: DayPlan[], travelers: number, rooms: number): number {
  return rounded(days.reduce((total, day) => total
    + day.attractions.reduce((sum, attraction) => sum + Math.max(0, Number(attraction.ticket_price ?? 0)) * travelers, 0)
    + day.meals.reduce((sum, meal) => sum + Math.max(0, Number(meal.estimated_cost ?? 0)) * travelers, 0)
    + (day.hotel && String(day.hotel.price_status ?? "") !== "unavailable"
      ? Math.max(0, Number(day.hotel.estimated_cost ?? 0)) * rooms
      : 0), 0));
}

function planPendingBuffer(days: DayPlan[], travelers: number, rooms: number): number {
  const transportRows = days.filter((day) => day.transportation.trim()).length;
  const unavailableHotels = days.filter((day) => day.hotel && day.hotel.price_status === "unavailable").length;
  return rounded(transportRows * TRANSPORT_BUFFER_PER_PERSON * travelers
    + unavailableHotels * HOTEL_BUFFER_PER_ROOM_NIGHT * rooms);
}

function adjustmentNote(language: unknown, reduction: number, buffer: number): string {
  const code = String(language ?? "zh").trim().toLocaleLowerCase("und").split("-")[0];
  if (code === "en") return `Meal estimates were reduced by ¥${reduction} to respect the budget, with ¥${buffer} reserved for unquoted items.`;
  if (code === "fr") return `Les estimations des repas ont été réduites de ¥${reduction} pour respecter le budget, avec ¥${buffer} réservés aux éléments sans tarif.`;
  return `已根据总预算将餐饮预估下调 ¥${reduction}，并为待报价项目预留 ¥${buffer}。`;
}

export function adjustGeneratedDaysToBudget(request: TripPlanningRequest, sourceDays: DayPlan[]): {
  days: DayPlan[];
  budget_adjustment_applied: boolean;
  budget_adjustment_note: string;
} {
  const days = structuredClone(sourceDays);
  const travelers = positiveCount(request.traveler_count);
  const rooms = positiveCount(request.room_count, Math.ceil(travelers / 2));
  const limit = groupBudgetLimit(request.budget_amount, request.budget_basis, travelers);
  const buffer = planPendingBuffer(days, travelers, rooms);
  const originalTotal = knownTotal(days, travelers, rooms);
  if (limit === null || originalTotal + buffer <= limit) {
    return { days, budget_adjustment_applied: false, budget_adjustment_note: "" };
  }

  let remaining = originalTotal + buffer - limit;
  const meals = days.flatMap((day) => day.meals).sort((left, right) => {
    const leftCost = Number(left.estimated_cost ?? 0);
    const rightCost = Number(right.estimated_cost ?? 0);
    const leftFloor = MEAL_FLOOR_PER_PERSON[left.type.toLocaleLowerCase("und")] ?? 20;
    const rightFloor = MEAL_FLOOR_PER_PERSON[right.type.toLocaleLowerCase("und")] ?? 20;
    return (rightCost - rightFloor) - (leftCost - leftFloor) || rightCost - leftCost;
  });
  for (const meal of meals) {
    if (remaining <= 0) break;
    const floor = MEAL_FLOOR_PER_PERSON[meal.type.toLocaleLowerCase("und")] ?? 20;
    const current = Number(meal.estimated_cost ?? 0);
    const reducible = Math.max(0, current - floor);
    const reductionPerPerson = Math.min(reducible, Math.ceil(remaining / travelers));
    if (reductionPerPerson <= 0) continue;
    meal.estimated_cost = current - reductionPerPerson;
    remaining = Math.max(0, remaining - reductionPerPerson * travelers);
  }
  const reduction = rounded(Math.max(0, originalTotal - knownTotal(days, travelers, rooms)));
  const applied = reduction > 0;
  return {
    days,
    budget_adjustment_applied: applied,
    budget_adjustment_note: applied ? adjustmentNote(request.language, reduction, buffer) : "",
  };
}
